File: day22/DayTwentyTwo.java
import java.util.Queue;
import java.util.LinkedList;

public class DayTwentyTwo
{
	class Node
	{
		int data;
		Node left;
		Node right;
		private Node(int data)
		{
			this.data = data;
			this.left = null;
			this.right = null;
		}
	}

	private int itr = 0;
	private int arr[] = {7,5,10,3,6,9,12};

	private Node btree = null;
	
	private Node createBtree()
	{
		for(int itr=0;itr<arr.length;itr++)
		{
			Node newnode = new Node(arr[itr]);
			btree = insertNode(btree,newnode);
		}
		return btree;
	}

	private Node insertNode(Node btree,Node newnode)
	{
		if(btree == null)
		{
			return newnode;
		}
		else if(newnode.data < btree.data)
		{
			btree.left = insertNode(btree.left,newnode);
		}
		else if(newnode.data > btree.data)
		{
			btree.right = insertNode(btree.right,newnode);
		}
		else if(newnode.data == btree.data)
		{
			System.out.print("Data Already Present");
			return btree;
		}
		return btree;
	}

	private void inorder(Node newnode)
	{
		if(newnode != null)
		{
			inorder(newnode.left);
			System.out.print(newnode.data + " ");
			inorder(newnode.right);
		}
	}


	private void preorder(Node newnode)
	{
		if(newnode != null)
		{
			System.out.print(newnode.data + " ");
			preorder(newnode.left);
			preorder(newnode.right);
		}
	}

	private void postorder(Node newnode)
	{
		if(newnode != null)
		{
			postorder(newnode.left);
			postorder(newnode.right);
			System.out.print(newnode.data + " ");
		}
	}

	private void levelorderTraversal(Node newnode)
	{
		Queue<Node> q = new LinkedList<Node>();
		q.add(newnode);
		while(q.size() != 0)
		{
			Node curr = q.peek();
			System.out.print(curr.data + " ");
			q.remove();
			if(curr.left != null)
			{
				q.add(curr.left);
			}

			if(curr.right != null)
			{
				q.add(curr.right);
			}
		}
	}

	public static void main(String args[])
	{
		DayTwentyTwo tree = new DayTwentyTwo();
		Node btree = tree.createBtree();
		System.out.print("Inorder Traversal of given Tree => ");
		tree.inorder(btree);
		System.out.println();
		System.out.print("Preorder Traversal of given Tree => ");
		tree.preorder(btree);
		System.out.println();
		System.out.print("Postorder Traversal of given Tree => ");
		tree.postorder(btree);
		System.out.println();
		System.out.print("levelorderTraversal of given Tree => ");
		tree.levelorderTraversal(btree);
		System.out.println();
	}
}

File: day22/day_twentytwo.cpp
# include <iostream>
# include <queue>

using namespace std;

typedef struct Node
{
	int data;
	struct Node* left;
	struct Node* right;
}node;

int itr = 0;
int arr[] = {7,5,10,3,6,9,12};
int size = sizeof(arr)/sizeof(arr[0]);


node* getNode()
{
	node* newnode = (node*)malloc(sizeof(node));
	return newnode;
}

void readNode(node* newnode)
{
	newnode->data = arr[itr];
	itr = itr + 1;
	newnode->left = NULL;
	newnode->right = NULL;
}

node* insertNode(node* btree,node* newnode)
{
	if(btree == NULL)
	{
		return newnode;
	}
	else if(newnode->data < btree->data)
	{
		btree->left = insertNode(btree->left,newnode);
	}
	else if(newnode->data > btree->data)
	{
		btree->right = insertNode(btree->right,newnode);
	}
	else if(btree->data == newnode->data)
	{
		cout << "data already exist" << endl;
		return btree;
	}
	return btree;
}

node* createBtree()
{
	node* btree = NULL;
	for(int i=0;i<size;i++)
	{
		node* newnode = getNode();
		readNode(newnode);
		btree = insertNode(btree,newnode);
	}
	return btree;
}

void inorder(node* newnode)
{
	if(newnode != NULL)
	{
		inorder(newnode->left);
		cout << newnode->data << " ";
		inorder(newnode->right);	
	}
}

void preorder(node* newnode)
{
	if(newnode != NULL)
	{
		cout << newnode->data << " ";
		preorder(newnode->left);
		preorder(newnode->right);
	}
}


void postorder(node* newnode)
{
	if(newnode != NULL)
	{
		postorder(newnode->left);
		postorder(newnode->right);
		cout << newnode->data << " ";
	}
}

void levelorder(node* newnode)
{
	queue<node*> q;
	q.push(newnode);
	while(q.empty() == false)
	{
		node* curr = q.front();
		cout << curr->data << " ";
		q.pop();
		if(curr->left != NULL)
		{
			q.push(curr->left);			
		}
		if(curr->right != NULL)
		{
			q.push(curr->right);			
		}
	}
}

int main()
{
	node* btree = createBtree();
	cout << "inorder Traversal => ";
	inorder(btree);
	cout << endl;
	cout << "preorder Traversal => ";
	preorder(btree);
	cout << endl;
	cout << "postorder Traversal => ";
	postorder(btree);
	cout << endl;
	cout << "levelorder Traversal => ";
	levelorder(btree);
	cout << endl;
	return 0;
}






